### PR TITLE
stm32/mboot: Support mass erase dfu command

### PR DESCRIPTION
**Pull request by andrewleech**
The mass erase command is currently not implemented in mboot due to the desire to not wipe the bootloader itself.

This PR adds a 'fake' mass erase command which erased sector-by-sector everything except the first mboot sector.

This change was originally included as part of https://github.com/micropython/micropython/pull/5309 but I've split it out as it grew into a significantly larger change-set than originally anticipated.

The erase command can take some time, on my stm32f765 with 2MB of flash it takes 8 to 10 seconds. This is normally enough to make pydfu fail on a timeout.

The DFU standard includes a mechanism for the dfu device to request a longer timeout as part of the get-status response just before starting an operation. This timeout functionality has been implemented here.

This PR required / is sitting on top of https://github.com/micropython/micropython/pull/5724 and https://github.com/micropython/micropython/pull/5728 github doesn't have and PR stacking yet so  these PR's commits are at the bottom of this one. Only the last two commits are relevant directly to this.

### CODE REVIEW

**Comment by andrewleech** [comment]
Build failure appears to be an intermittent(?) issue with the esp32 build unrelated to this PR

**Comment by dpgeorge** [comment]
Adding mass erase is a good feature.  One reason I never got to it was because it was hard to decide what to do about external SPI flash: should it also be erased?  (I guess it's not in this PR.)  Eg with XIP flash it's basically an extension of internal flash, so if you are asking for a full erase then you'd probably also want the XIP to be erased.

But what exactly is the need for mass erase in your case?  I guess it'd usually be because it's faster to mass erase then write, rather than erasing sector-by-sector.  But here that won't make a difference due to the "fake" mass erase.

**Comment by andrewleech** [comment]
My specific use case it upgrading in-field to my new signed bootloader, once mboot is flashed (by application code), the board reboots into it with `machine.bootloader()`. The first thing I do then is issue a mass erase from pydfu to stop the board from booting back into the old application in between flashing the encryption key footer and main application.

A page erase function could be exposed and used in pydfu, but then the flashing scripts needs to know which page by sector number or address needs to be erased, which is far messier than a single mass erase command.

Yeah external flash is a complication I hadn't considered, certainly not in an XIP / extended mode. I was basically just trying to replicate what the rom stm dfu did as closely as was sensible.

**Comment by dpgeorge** [comment]
@andrewleech when you get a chance can you please rebase this on latest master, then I can review it properly.

**Comment by andrewleech** [comment]
@dpgeorge Thanks for the cleanups / changes on the merged dependent branches, they looked good.

I've rebased this and believe it looks good to go.

File: ports/stm32/mboot/main.c [hunk 1]
```diff
@@ -964,6 +975,35 @@ static void dfu_init(void) {
...
+                return flash_length / 200;
+
+            } else if (dfu_context.wLength == 5) {
+                // erase page command
+                return 500;
+            }
+        } else if (dfu_context.wLength >= 1 && dfu_context.buf[0] == 0x21) {
+            if (dfu_context.wLength == 5) {
+                // set address command
+                return 0;
```
Comment (peitschie) [inline ports/stm32/mboot/main.c:997]: This case could go away, as the function returns 0 by default.

File: ports/stm32/mboot/main.c [hunk 2]
```diff
@@ -972,6 +1012,11 @@ static int dfu_process_dnload(void) {
             if (dfu_context.wLength == 1) {
                 // mass erase
                 ret = do_mass_erase();
+                if (ret == -1) {
+                    dfu_context.cmd = DFU_CMD_NONE;
+                    return DFU_STATE_ERROR;
```
Comment (peitschie) [inline ports/stm32/mboot/main.c:1017]: Just noting the early return here is inconsistent with the rest of this function, which uses the `ret` value instead and just lets execution fall to the bottom.

File: ports/stm32/mboot/main.c [hunk 3]
```diff
@@ -972,6 +1012,11 @@ static int dfu_process_dnload(void) {
...
+                }
+                return DFU_STATE_DNLOAD_IDLE;
```
Comment (peitschie) [inline ports/stm32/mboot/main.c:1019]: Same thing here.

**Review by peitschie — COMMENTED** [review]
Just noting a few things I saw in a quick fly-over! ☺️

File: ports/stm32/mboot/main.c [hunk 4]
```diff
@@ -97,7 +97,7 @@ void mp_hal_delay_us(mp_uint_t usec) {
     }
 }
 
-static volatile uint32_t systick_ms;
+static volatile uint32_t systick_ms = 0;
```
Comment (dpgeorge) [inline ports/stm32/mboot/main.c:100]: is this change needed?  it should already be in bss

File: ports/stm32/mboot/main.c [hunk 5]
```diff
@@ -562,7 +562,7 @@ static inline bool flash_is_valid_addr(uint32_t addr) {
     return flash_layout[0].base_address <= addr && addr < end_of_flash;
 }
 
-static uint32_t flash_get_sector_index(uint32_t addr, uint32_t *sector_size) {
+static int32_t flash_get_sector_index(uint32_t addr, uint32_t *sector_size, uint32_t *sector_start) {
```
Comment (dpgeorge) [inline ports/stm32/mboot/main.c:565]: this function should now be available from `stm32/flash.h`

File: ports/stm32/mboot/main.c [hunk 6]
```diff
@@ -698,9 +700,18 @@ static int flash_write(uint32_t addr, const uint8_t *src8, size_t len) {
...
-    // TODO
-    return flash_mass_erase();
+static int do_mass_erase() {
+    uint32_t sector_num = 0;
+    for (int i = 0; i < MP_ARRAY_SIZE(flash_layout); ++i) {
+        for (int j = 0; j < flash_layout[i].sector_count; ++j) {
+            sector_num ++;
+        }
+    }
+    if (flash_erase_hal(1, sector_num-1) != 0) {
```
Comment (dpgeorge) [inline ports/stm32/mboot/main.c:710]: note that some MCUs (eg WB55) have mboot in multiple sectors because the sector size is 4k, so starting at sector 1 won't work for them.  well, I think that's a bit out of scope of this PR, it can be fixed later (so this mass erase only works on F4/F7s for the time being)

File: ports/stm32/mboot/main.c [hunk 7]
```diff
@@ -464,31 +466,37 @@ static int usrbtn_state(void) {
...
+    #endif
 }
 
 static int mboot_flash_page_erase(uint32_t addr, uint32_t *next_addr) {
     uint32_t sector_size = 0;
-    uint32_t sector = flash_get_sector_info(addr, NULL, &sector_size);
-    if (sector == 0) {
-        // Don't allow to erase the sector with this bootloader in it
+    uint32_t sector_start = 0;
+    int32_t sector = flash_get_sector_info(addr, &sector_start, &sector_size);
```
Comment (dpgeorge) [inline ports/stm32/mboot/main.c:481]: `flash_get_sector_info` returns an unsigned value, and 0 for invalid address... I guess that function should be changed to return -1 on error

File: ports/stm32/mboot/main.c [hunk 8]
```diff
@@ -888,12 +921,16 @@ static int dfu_handle_tx(int cmd, int arg, int len, uint8_t *buf, int max_len) {
...
-        buf[5] = dfu_context.error; // iString
+        size_t timeout_ms = get_timeout_ms();
+        buf[0] = dfu_context.status;          // bStatus
+        buf[1] = (timeout_ms >> 16) & 0xFF;   // bwPollTimeout (ms)
+        buf[2] = (timeout_ms >> 8) & 0xFF;    // bwPollTimeout (ms)
+        buf[3] = timeout_ms & 0xFF;           // bwPollTimeout (ms)
+        buf[4] = dfu_context.state;           // bState
+        buf[5] = dfu_context.error;           // iString
+        // Clear errors now they've been sent
+        dfu_context.status = 0;
```
Comment (dpgeorge) [inline ports/stm32/mboot/main.c:932]: `DFU_STATUS_OK`?

File: ports/stm32/mboot/main.c [hunk 9]
```diff
@@ -441,6 +441,8 @@ static int usrbtn_state(void) {
 /******************************************************************************/
 // FLASH
 
+#define APPLICATION_FLASH_LENGTH (FLASH_END + 1 - APPLICATION_ADDR)
```
Comment (dpgeorge) [inline ports/stm32/mboot/main.c:444]: for now, to get this building (and working) on WB55 I think put the following above this line:
```c
#if defined(STM32WB)
#define FLASH_END FLASH_END_ADDR
#endif 
```

File: ports/stm32/mboot/main.c [hunk 10]
```diff
@@ -464,31 +466,37 @@ static int usrbtn_state(void) {
 #endif
 
 static int mboot_flash_mass_erase(void) {
-    // TODO
+    #if defined(STM32F4) || defined(STM32F7) || defined(STM32H7)
```
Comment (dpgeorge) [inline ports/stm32/mboot/main.c:469]: with the above fix for `APPLICATION_FLASH_LENGTH` this #if guard should be able to be removed

File: ports/stm32/mboot/main.c (hunk 7 above)
Comment (andrewleech) [reply ports/stm32/mboot/main.c:481]: Ah yes, I'd made a similar change to the equivalent local copy of that function previously.

**Comment by dpgeorge** [comment]
This has been rebased and merged in 4d6f60d4286b3e6ea8c9fb0141ba13e4978882b6 through 40006813c3e053e4cf952bfdd32866ce9496b0f8
